packages/core/src/agent/workflow/function-agent.ts

```diff
@@ -0,0 +1,176 @@
+  private logger: {
+    log: (...args: unknown[]) => void;
+    error: (...args: unknown[]) => void;
+    warn: (...args: unknown[]) => void;
+    info: (...args: unknown[]) => void;
+  };
+
+  constructor(params: {
+    name: string;
+    llm: LLM;
+    scratchpadKey?: string;
+    verbose?: boolean;
+  }) {
+    this.name = params.name;
+    this.llm = params.llm;
+    this.logger = params.verbose ? consoleLogger : emptyLogger;
+    if (params.scratchpadKey) {
+      this.scratchpadKey = params.scratchpadKey;
+    }
+  }
+
+  async takeStep(
+    ctx: HandlerContext<AgentWorkflowContext>,
+    tools: BaseToolWithCall[],
+    memory: BaseMemory,
+  ): Promise<AgentOutput> {
+    // Get scratchpad from context or initialize if not present
+    const scratchpad: ChatMessage[] = ctx.data.scratchpad;
+
+    const memoryMessages = await memory.getMessages();
+
+    this.logger.info(
+      `Calling LLM with messages: ${JSON.stringify(memoryMessages)}`,
+    );
+    const response = await this.llm.chat({
+      messages: memoryMessages,
+      tools,
+      stream: false,
+    });
+
+    // Get tool calls from response
+    const toolCalls: ToolCall[] = [];
+    const options = response.message.options ?? {};
+
+    // Extract tool calls from the response if they exist
+    if (options && "toolCall" in options && Array.isArray(options.toolCall)) {
+      this.logger.info(
+        `Found ${options.toolCall.length} tool calls in response`,
+      );
+
+      for (const call of options.toolCall) {
+        try {
+          // Convert input to arguments format
+          let args: Record<string, unknown>;
+          if (typeof call.input === "string") {
+            try {
+              args = JSON.parse(call.input);
+            } catch (e) {
+              args = { rawInput: call.input };
+            }
+          } else {
+            args = call.input as Record<string, unknown>;
+          }
+
+          toolCalls.push({
+            id: call.id,
+            name: call.name,
+            input: args as JSONObject,
+          });
+        } catch (error) {
+          this.logger.error(`Error processing tool call: ${error}`);
+        }
+      }
+    }
+
+    // Add response to scratchpad
+    scratchpad.push(response.message);
+    ctx.data.scratchpad = scratchpad;
+
+    return {
+      response: response.message,
+      toolCalls,
+      raw: response.raw,
+      currentAgentName: this.name,
+    };
+  }
+
+  async handleToolCallResults(
+    ctx: HandlerContext<AgentWorkflowContext>,
+    results: ToolCallResult[],
+    memory: BaseMemory,
+  ): Promise<void> {
+    const scratchpad: ChatMessage[] = ctx.data.scratchpad;
+
+    // Process each tool result and add to scratchpad only (not memory yet)
+    for (const result of results) {
+      // Format the output content
+      const content = formatToolOutput(result.toolOutput.output);
+
+      const rawToolMessage = {
+        role: "assistant" as MessageType,
```

Comment:
try to avoid as, probably this works:
```suggestion
      const rawToolMessage: Message = {
        role: "assistant",
```